src/ray/core_worker/core_worker.cc

```diff
@@ -1242,121 +1245,22 @@ void CoreWorker::RemoveActorHandleReference(const ActorID &actor_id) {
-      }
-
-      const auto &actor_state = gcs::ActorTableData::ActorState_Name(actor_data.state());
-      RAY_LOG(INFO) << "received notification on actor, state: " << actor_state
-                    << ", actor_id: " << actor_id
-                    << ", ip address: " << actor_data.address().ip_address()
-                    << ", port: " << actor_data.address().port() << ", worker_id: "
-                    << WorkerID::FromBinary(actor_data.address().worker_id())
-                    << ", raylet_id: "
-                    << ClientID::FromBinary(actor_data.address().raylet_id());
-    };
-
-    RAY_CHECK_OK(gcs_client_->Actors().AsyncSubscribe(
-        actor_id, actor_notification_callback, nullptr));
-
-    RAY_CHECK(reference_counter_->SetDeleteCallback(
-        actor_creation_return_id,
-        [this, actor_id, is_owner_handle](const ObjectID &object_id) {
-          if (is_owner_handle) {
-            // If we own the actor and the actor handle is no longer in scope,
-            // terminate the actor. We do not do this if the GCS service is
-            // enabled since then the GCS will terminate the actor for us.
-            if (!RayConfig::instance().gcs_actor_service_enabled()) {
-              RAY_LOG(INFO) << "Owner's handle and creation ID " << object_id
-                            << " has gone out of scope, sending message to actor "
-                            << actor_id << " to do a clean exit.";
-              RAY_CHECK_OK(
-                  KillActor(actor_id, /*force_kill=*/false, /*no_restart=*/false));
-            }
-          }
-
-          absl::MutexLock lock(&actor_handles_mutex_);
-          // TODO(swang): Erase the actor handle once all refs to the actor
-          // have gone out of scope. We cannot erase it here in case the
-          // language frontend receives another ref to the same actor. In this
-          // case, we must remember the last task counter that we sent to the
-          // actor.
-          // TODO(ekl) we can't unsubscribe to actor notifications here due to
-          // https://github.com/ray-project/ray/pull/6885
-          auto callback = actor_out_of_scope_callbacks_.extract(actor_id);
-          if (callback) {
-            callback.mapped()(actor_id);
-          }
-        }));
-  }
-
-  return inserted;
+  std::unique_ptr<ActorHandle> &actor_handle = actor_manager_->GetActorHandle(actor_id);
+  actor_handle->Serialize(output);
+  *actor_handle_id = ObjectID::ForActorHandle(actor_id);
+  return Status::OK();
 }
 
-Status CoreWorker::GetActorHandle(const ActorID &actor_id,
-                                  ActorHandle **actor_handle) const {
-  absl::MutexLock lock(&actor_handles_mutex_);
-  auto it = actor_handles_.find(actor_id);
-  if (it == actor_handles_.end()) {
-    return Status::Invalid("Handle for actor does not exist");
-  }
-  *actor_handle = it->second.get();
-  return Status::OK();
+void CoreWorker::GetActorHandle(const ActorID &actor_id,
```

Comment:
I can consider making it return unique_ptr<ActorHandle> instead. I just kept the current structure because the language frontend doesn't really use this now.

Reply:
How about we only return the values in the actor handle?

Reply:
(So we don't have to return a raw pointer here)

Reply:
What about we just return the const pointer? I believe the pointed value cannot be modified by the const pointer. Are you concerning about this raw pointer being stored somewhere in the language frontend? 

It seems like returning values here would require us to use at least one copy operations because we need to 
1. copy from cpp -> cpp struct
2. copy from cpp struct -> language frontend.

I am a little concerned about that `extension_data` could be expensive to copy twice. (idk which data is used for this, but it seems like it will be some json string?)